### PR TITLE
fix(ec2): "clientVpnEndoint" => "clientVpnEndpoint"

### PR DESCRIPTION
**Pull request by cokia**
It is a solution to the issue #13810

Corrected the typo "clientVpnEndpoint" in the ec2 module to "clientVpnEnpoint".

The maintainer (@rix0rrr) talked about @deprecated tags on existing incorrect members in the issue thread, but I didn't work on it because I thought it was an internal reference member. (if @deprecated tag is the rule of this repository in this case, please comment).

This is my first commitment to this report. If there's anything wrong, please let me know in the comments. thanks

---- edit -----
Originally, this PR was #14835 , but I deleted my fork report by mistake, so I couldn't correct the error, so I uploaded it again.

### CODE REVIEW

**Comment by cokia** [comment]
When I tested locally, it was built normally and even tested, but I don't know why the server got such an error.

**Comment by rix0rrr** [comment]
> When I tested local, it was built normally and even tested, but I don't know why the server got such an error.

This is the error:

```
API elements with incompatible changes:
err  - PROP @aws-cdk/aws-ec2.ClientVpnAuthorizationRuleProps.clientVpnEndoint: has been removed [removed:@aws-cdk/aws-ec2.ClientVpnAuthorizationRuleProps.clientVpnEndoint]
err  - IFACE @aws-cdk/aws-ec2.ClientVpnAuthorizationRuleProps: newly required property 'clientVpnEndpoint' added: input to @aws-cdk/aws-ec2.ClientVpnAuthorizationRule.<initializer> [strengthened:@aws-cdk/aws-ec2.ClientVpnAuthorizationRuleProps]
err  - PROP @aws-cdk/aws-ec2.ClientVpnRouteProps.clientVpnEndoint: has been removed [removed:@aws-cdk/aws-ec2.ClientVpnRouteProps.clientVpnEndoint]
err  - IFACE @aws-cdk/aws-ec2.ClientVpnRouteProps: newly required property 'clientVpnEndpoint' added: input to @aws-cdk/aws-ec2.ClientVpnRoute.<initializer> [strengthened:@aws-cdk/aws-ec2.ClientVpnRouteProps]
```

I think you saw this coming when you said: 

> The maintainer (@rix0rrr) talked about @deprecated tags on existing incorrect members in the issue thread, but I didn't work on it because I thought it was an internal reference member.

😀

**Comment by rix0rrr** [comment]
(Bottom line, we are not allowed to break existing code -- so we're going to have to go the deprecated+alternative route)

**Review by rix0rrr — CHANGES_REQUESTED** [review]
We need to go the non-breaking route

**Review by cokia — COMMENTED** [review]
I added deprecated tag to typo route

**Comment by cokia** [comment]
Linter did not allow me to delete the typo type, and I made it an option type and added a new one to solve this problem. Is this the right solution? I don't know if it's the right choice to manage this optionally defined type.

**Comment by rix0rrr** [comment]
> I made it an optional type and added a new one to solve this problem

This is exactly the right way. Plus adding validation (in code) to make sure exactly one of the parameters is supplied (either the typoed or non-typoed one).

File: packages/@aws-cdk/aws-ec2/lib/client-vpn-authorization-rule.ts [hunk 1]
```diff
@@ -36,7 +36,14 @@ export interface ClientVpnAuthorizationRuleProps extends ClientVpnAuthorizationR
   /**
    * The client VPN endpoint to which to add the rule.
    */
-  readonly clientVpnEndoint: IClientVpnEndpoint;
+  readonly clientVpnEndpoint: IClientVpnEndpoint;
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/client-vpn-authorization-rule.ts:39]: Aha, here is the confusion. Both need to be declared optional (type-wise), and runtime code needs to validate exactly one of them is supplied anyway.

Don't forget that both need a `@default` annotation as well. I'd recommend something like:

```
@default clientVpnEndpoint is required
```

**Review by rix0rrr — CHANGES_REQUESTED** [review]

**Comment by cokia** [comment]
@rix0rrr build succeeded!
please recheck this PR 🚀

**Comment by cokia** [comment]
I'm sorry to tag you again.
If you don't mind, could you recheck this PR? @rix0rrr

**Comment by douglasnaphas** [comment]
@cokia You still have a [linter failure](https://github.com/aws/aws-cdk/pull/14902/checks?check_run_id=2948626370), with the message:

> Error: Fixes must contain a change to a test file

@rix0rrr recommended [above](https://github.com/aws/aws-cdk/pull/14902#issuecomment-860676290):

> ...adding validation (in code) to make sure exactly one of the parameters is supplied (either the typoed or non-typoed one).

I suggest adding the following tests:

1. Instantiate a `ClientVpnAuthorizationRule`, supplying `clientVpnEndpoint` and `clientVpnEndoint`. Expect an error explaining that one or the other is required, but both are not allowed.
2. Instantiate a `ClientVpnAuthorizationRule`, supplying neither `clientVpnEndpoint` nor `clientVpnEndoint`. Expect an error explaining that one or the other is required.
3. Instantiate a `ClientVpnRoute`, supplying `clientVpnEndpoint` and `clientVpnEndoint`. Expect an error explaining that one or the other is required, but both are not allowed.
4. Instantiate a `ClientVpnRoute`, supplying neither `clientVpnEndpoint` nor `clientVpnEndoint`. Expect an error explaining that one or the other is required.

Commit the tests, get builds that fail because the logic isn't in place in the code yet, then implement checks that will make the tests pass in [ClientVpnAuthorizationRule](https://github.com/aws/aws-cdk/blob/3e63d2840fdd9862e3fff5466646001b2713721b/packages/%40aws-cdk/aws-ec2/lib/client-vpn-authorization-rule.ts#L53) and [ClientVpnRoute](https://github.com/aws/aws-cdk/blob/3e63d2840fdd9862e3fff5466646001b2713721b/packages/%40aws-cdk/aws-ec2/lib/client-vpn-route.ts#L83).

Also I don't think the `!` on [this line](https://github.com/aws/aws-cdk/blob/3e63d2840fdd9862e3fff5466646001b2713721b/packages/%40aws-cdk/aws-ec2/lib/client-vpn-authorization-rule.ts#L58)

```ts
      clientVpnEndpointId: props.clientVpnEndpoint!.endpointId,
```

and [this identical line](https://github.com/aws/aws-cdk/blob/3e63d2840fdd9862e3fff5466646001b2713721b/packages/%40aws-cdk/aws-ec2/lib/client-vpn-route.ts#L87):

```ts
      clientVpnEndpointId: props.clientVpnEndpoint!.endpointId,
```

can stay, because we want to allow for existing usages of `ClientVpnAuthorizationRuleProps` and `ClientVpnRouteProps` that  use `clientVpnEndoint` to still work.

**Comment by douglasnaphas** [comment]
@cokia As noted in [my earlier comment](https://github.com/aws/aws-cdk/pull/14902#issuecomment-873609825), this PR needs updates to tests, and some changes to the logic to enforce the requirement that one or the other or `clientVpnEndoint` (typo) or `clientVpnEndpoint` is provided.

I pulled down your branch, and made commits that I believe will get this PR passing the linter and able to pass review.

I have my changes on my branch [douglasnaphas:fix-clientVpnEndpoint-doug](https://github.com/aws/aws-cdk/compare/master...douglasnaphas:fix-clientVpnEndpoint-doug?expand=1). Take a look at that, and if you agree with those changes, you could:
  1. Merge `aws/aws-cdk:master` into your PR's branch.
  2. Merge [douglasnaphas/aws-cdk:fix-clientVpnEndpoint-doug](https://github.com/douglasnaphas/aws-cdk/tree/fix-clientVpnEndpoint-doug) into your your PR's branch.

Then I think we will be in good shape for the maintainers to take a look and get this change merged.

**Comment by cokia** [comment]
@douglasnaphas  I am not familiar with the structure of this project and the basics for the contribution. You've helped me a lot, and I'm very grateful for this.

**Comment by douglasnaphas** [comment]
@rix0rrr, the build that was previously failing on this PR is passing, so it is ready for another review when you have a chance.

**Review by rix0rrr — APPROVED** [review]

**Pull request merged**